nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+        lengths: Tensor,
+        zs: Optional[List[Tensor]] = None,
+        ws: Optional[List[Tensor]] = None,
+        noise: Optional[Tensor] = None,
+        mixing: bool = False,
+        normalize: bool = True,
+    ):
+        if len(condition.shape) != 4:
+            raise ValueError(f"Got spectrogram tensor of shape {condition.shape}, expected B x 1 x C x L")
+
+        batch_size, *_, max_length = condition.shape
+
+        # generate noise
+        if zs is not None and ws is not None:
+            raise ValueError(
+                "Please specify either zs or ws or neither, but not both. It is not clear which one to use."
+            )
+
+        if zs is None:
+            zs = self.generate_zs(batch_size, mixing)
+        if ws is None:
+            ws = [self.G.style_mapping(z) for z in zs]
+        if noise is None:
+            noise = self.generate_noise(batch_size)
+
+        # normalize if needed, mask and pad appropriately
+        if normalize:
+            condition = self.normalize_spectrograms(condition)
+        condition = self.pad_spectrogram(condition)
+        condition = mask(condition, lengths)
+
+        # the main call
+        enhanced_spectrograms = self.G(condition, lengths, ws, noise)
+
+        # denormalize if needed, mask and remove padding
+        if normalize:
+            enhanced_spectrograms = self.unnormalize_spectrograms(enhanced_spectrograms)
+        enhanced_spectrograms = mask(enhanced_spectrograms, lengths)
+        enhanced_spectrograms = enhanced_spectrograms[:, :, :, :max_length]
+
+        return enhanced_spectrograms
+
+    def prepare_batch(self, batch):
+        if not isinstance(self.spectrogram_model, FastPitchModel):
+            raise NotImplementedError(
+                f"{type(self.spectrogram_model)} is not supported, please implement batch preparation for this model."
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
+        mels, mel_lens = self.spectrogram_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        mels_pred, *_, pitch = self.spectrogram_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=mel_lens,
+            input_lens=text_lens,
+        )
+
+        mels = mels.cpu()
+        mel_lens = mel_lens.cpu()
+        mels_pred = mels_pred.cpu()
+
+        batch_size, *_ = mels.shape
+
+        targets = []
+        conditions = []
+        lengths = []
+
+        for i in range(batch_size):
+            l = mel_lens[i].item()
+            lengths.append(l)
+
+            targets.append(mels[i, :, :l].T)
+            conditions.append(mels_pred[i, :, :l].T)
+
+        target = nn.utils.rnn.pad_sequence(targets, batch_first=True)
```

Comment:
Best to keep these kind of padding and data manipulation logic in dataloader.

Reply:
I initially wrote it as a separate dataloader and gave up after some debugging. 
In this implementation data loading is delegated to underlying spectrogram generator. Which ended up much more robust and clean.

Should I move it to its own class? Still would be a regular class, not a torch/nemo dataset or dataloader or collate_fn

Reply:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L995
Take a look at this part, it can help with padding.

Reply:
Sorry, I don't understand your last message about padding, what is your suggestion or what is the issue.

Reply:
Thanks Ryan, both are good points 

I guess at the initial research stage I copied messy code from the precomputing spectrograms script and reassembled tensors just to be absolute sure about data format. These were B x 1 x D x L for 2d convs.

Will remove redundancies

Reply:
Done but will be replaced with a preprocessing step